src/main/java/net/minecraftforge/event/terraingen/DecorateBiomeEvent.java

```diff
@@ -96,27 +98,94 @@ public Post(World world, Random rand, BlockPos pos)
 
     /**
      * This event is fired when a chunk is decorated with a biome feature.
-     *
+     * The amount denotes how many instances of the feature will be generated.
+     * <p>
      * You can set the result to DENY to prevent the default biome decoration.
      */
     @HasResult
     public static class Decorate extends DecorateBiomeEvent
     {
+        private final EventType type;
+        @Nullable
+        private final Integer totalAmount;
+        @Nullable
+        private Integer modifiedAmount;
+
```

Comment:
Easy. You set your event priority and whoever gets called first gets to decide first. Vanilla is 10, B wants 2 so it sets to 2, C wants cancel so it cancels. C is given the opportunity to check if the amount has been changed or the event canceled already and react accordingly.